src/subscription_manager/scripts/rhsmcertd_worker.py

```diff
@@ -92,78 +99,165 @@ def _collect_cloud_info(cloud_list: list) -> dict:
-    # included in JWT token
-    try:
-        register_service.register(org=None, jwt_token=jwt_token)
-    except Exception as err:
-        logger.error("Unable to auto-register: {err}".format(err=err))
-        sys.exit(-1)
+    logger.error(f"Unsupported token type for automatic registration: {token['tokenType']}.")
+    sys.exit(1)
+
+
+def _auto_register_standard(uep: "UEPConnection", token: Dict[str, str]) -> None:
+    """Perform standard automatic registration.
+
+    :raises Exception: The system could not be registered.
+    """
+    logger.debug("Registering the system through standard automatic registration.")
+
+    service = RegisterService(cp=uep)
+    service.register(org=None, jwt_token=token["token"])
+
+
+def _auto_register_anonymous(uep: "UEPConnection", token: Dict[str, str]) -> None:
+    """Perform anonymous automatic registration.
+
+    First we download the anonymous entitlement certificates and install them.
+
+    Then we wait the 'splay' period. This makes sure we give the cloud backend
+    enough time to create all the various objects in the Candlepin database.
+
+    Then we perform the registration to obtain identity certificate and proper
+    entitlement certificates.
+
+    :raises Exception: The system could not be registered.
+    """
+    logger.debug("Registering the system through anonymous automatic registration.")
+
+    # Step 1: Get the anonymous entitlement certificates
+    manager = entcertlib.AnonymousCertificateManager(uep=uep)
+    manager.register(uuid=token["anonymousConsumerUuid"], jwt=token["token"])
+
+    # Step 2: Wait
+    cfg = config.get_config_parser()
+    if cfg.get("rhsmcertd", "splay") == "0":
+        logger.debug("Trying to obtain the identity immediately, splay is disabled.")
     else:
-        logger.debug("Auto-registration performed successfully")
-        sys.exit(0)
+        registration_interval = int(cfg.get("rhsmcertd", "auto_registration_interval"))
+        splay_interval: int = random.randint(60, registration_interval * 60)
+        logger.debug(
+            f"Waiting a period of {splay_interval} seconds "
+            f"(about {splay_interval // 60} minutes) before attempting to obtain the identity."
+        )
+        time.sleep(splay_interval)
+
+    # Step 3: Obtain the identity certificate
+    logger.debug("Obtaining system identity")
+
+    registration_attempt: int = 0
+    service = RegisterService(cp=uep)
+    while True:
+        # If the server does not have the identity prepared yet, it returns
+        # an error with status code 451 and a Retry-After header.
+        # As the Candlepin database gets filled with data, the Retry-After
+        # intervals get shorter and shorter.
+        #
+        # Because `while` loops with no definite exit conditions are scary and
+        # tend to go wrong, we limit this to ten attempts.
+        registration_attempt += 1
+        if registration_attempt > 10:
```

Comment:
Yeah, current Candlepin implementation uses `Retry-After` values of 300, 180, 60 and 30 seconds as it progresses. That is why I think ten attempts would be fine -- 300*10 is fifty minutes, so at worst case we have been trying for nearly an hour. If the value is shorter, it means we are getting closer to completion, and will likely not run into the limit.

An alternative still using counters could be to exit if we got the same `Retry-After` value 3-4 times in a row -- that would mean something got stuck on the server side and we should wait until the next autoregister cycle to run. Server could have fixed this in that time period.

Let's consider your approach: Since the `auto_registration_interval` from the config file is still being used, should the time-based `TimeoutError` be raised at e.g. 75 % of this interval? This would make sure we try as long as possible while more or less preventing the C wrapper from starting a second instance of the worker (could that cause problems?).

Any of these will do pretty much the same. We still don't know how the actual server will behave, we have to take a guess now and perhaps correct it in future if the guess is not the optimal solution.

Reply:
Just a note here: candlepin is sending HTTP 429 with the Retry-After, not HTTP 451

Reply:
Ugh, you are right, I have no idea why there is a different status code number.